raw/fetcher.go

```diff
@@ -0,0 +1,227 @@
+package raw
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"time"
+
+	agent "github.com/bolt-observer/agent/agent"
+	api "github.com/bolt-observer/agent/lightning"
+	backoff "github.com/cenkalti/backoff/v4"
+	"github.com/golang/glog"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/metadata"
+	"google.golang.org/grpc/status"
+)
+
+// Fetcher struct
+type Fetcher struct {
+	AuthToken    string
+	LightningAPI api.LightingAPICalls
+	AgentAPI     agent.AgentAPIClient
+	PubKey       string
+	ClientType   int
+}
+
+// MakeFetcher creates a new fetcher
+func MakeFetcher(authToken string, endpoint string, l api.LightingAPICalls) (*Fetcher, error) {
+	f := &Fetcher{
+		AuthToken:    authToken,
+		LightningAPI: l,
+	}
+
+	info, err := l.GetInfo(context.Background())
+	if err != nil {
+		return nil, err
+	}
+
+	typ := 0
+	switch l.GetAPIType() {
+	case api.LndGrpc:
+		typ = 0
+	case api.LndRest:
+		typ = 1
+	case api.ClnSocket:
+		typ = 2
+	}
+	f.ClientType = typ
+
+	f.PubKey = info.IdentityPubkey
+	agent := GetAgentAPI(endpoint, f.PubKey, f.AuthToken)
+	if agent == nil {
+		return nil, err
+	}
+	f.AgentAPI = agent
+
+	return f, nil
+}
+
+func makePermanent(err error) error {
+	st := status.Convert(err)
+	if st.Code() == codes.Unknown {
+		if strings.Contains(st.Message(), "ConditionalCheckFailedException") {
+			return backoff.Permanent(err)
+		}
+	}
+
+	return err
+}
+
+// FetchInvoices will fetch and report invoices
+func (f *Fetcher) FetchInvoices(ctx context.Context, updateTimeWithLast bool, from time.Time) {
+
+	ctx = metadata.AppendToOutgoingContext(ctx, "pubkey", f.PubKey, "clientType", fmt.Sprintf("%d", f.ClientType), "key", f.AuthToken)
+
+	if updateTimeWithLast {
+		ts, err := f.AgentAPI.LatestInvoiceTimestamp(ctx, &agent.Empty{})
+
+		if err != nil {
+			glog.Warningf("Coud not get latest invoice timestamps: %v", err)
+		}
+
+		if ts != nil {
+			t := time.Unix(0, ts.Timestamp)
+			glog.V(2).Infof("Latest invoice timestamp: %v", t)
+			if t.After(from) {
+				from = t
+			}
+		}
+	}
+
+	outchan := GetInvoicesChannel(ctx, f.LightningAPI, from)
+
+	num := 0
+
+	b := backoff.NewExponentialBackOff()
+	b.MaxElapsedTime = 0
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case invoice := <-outchan:
+			data := &agent.DataRequest{
+				Timestamp: invoice.Timestamp.UnixNano(),
+				Data:      string(invoice.Message),
+			}
+
+			backoff.RetryNotify(func() error {
```

Comment:
if we cannot send data we should stop this for loop. client must wait until backend is ready otherwise we might have gaps.

You are not handling error returned from RetryNotify

Reply:
Good catch